proc/put.go

```diff
@@ -70,153 +80,133 @@ func (p *Put) maybeWarn(err error) {
+		typ:         typ,
+		clauseTypes: clauseTypes,
+		replace:     replace,
+		append:      tail,
+	}, nil
+}
 
-		newinfo := descinfo{
-			typ:    typ,
-			fields: fields,
-			order:  order,
+func clauseTypesMatch(types []clauseType, vals []zng.Value) bool {
+	for k, typ := range types {
+		if vals[k].Type != typ.Type {
+			return false
 		}
+	}
+	return true
+}
 
-		p.outmap[in.Type.ID()] = newinfo
-		descInfo = newinfo
+func (p *Put) lookupRule(inType *zng.TypeRecord, vals []zng.Value) (*putRule, error) {
+	rule := p.rules[inType.ID()]
+	if rule != nil && clauseTypesMatch(rule.clauseTypes, vals) {
+		return rule, nil
 	}
+	var err error
+	rule, err = p.buildRule(inType, vals)
+	p.rules[inType.ID()] = rule
+	return rule, err
+}
 
-	// Build the new output value
+func (p *Put) put(in *zng.Record) *zng.Record {
+	vals, err := p.eval(in)
+	if err != nil {
+		p.maybeWarn(err)
+		return in
+	}
+	rule, err := p.lookupRule(in.Type, vals)
+	if err != nil {
+		p.maybeWarn(err)
+		return in
+	}
+	// Start the new output value by either copying or replacing the input values.
 	var bytes zcode.Bytes
-	i := 0
-	if descInfo.fields[descInfo.order[i]].position == -1 {
-		// all fields are being appended...
+	if rule.replace == nil {
+		// All fields are being appended.
 		bytes = in.Raw
 	} else {
-		// we're overwriting one or more fields.  descInfo.order
-		// has the column numbers we need, sorted.
-		origCols := len(in.Type.Columns)
+		// We're overwriting one or more fields.  Travese the
+		// replacement vector to determine whether each value should
+		// be copied from the input or replaced with a clause result.
 		iter := in.ZvalIter()
-		vali := descInfo.order[i]
-		nextcol := descInfo.fields[vali].position
-		for col := range descInfo.typ.Columns {
-			if col == origCols {
-				break
-			}
+		for _, clause := range rule.replace {
 			item, isContainer, err := iter.Next()
 			if err != nil {
 				panic(err)
 			}
-			if col == nextcol {
-				item = p.vals[vali].Bytes
-				isContainer = descInfo.fields[vali].container
-
-				// Advance to the next column we need to
-				// overwrite.  If we're done overwriting
-				// columns, just set nextcol to -1 so this
-				// loop finishes copying existing fields.
-				i++
-				if i < len(descInfo.fields) {
-					vali = descInfo.order[i]
-					nextcol = descInfo.fields[vali].position
-				} else {
-					nextcol = -1
-				}
+			if clause >= 0 {
+				item = vals[clause].Bytes
+				isContainer = rule.clauseTypes[clause].container
 			}
-
 			if isContainer {
 				bytes = zcode.AppendContainer(bytes, item)
 			} else {
 				bytes = zcode.AppendPrimitive(bytes, item)
 			}
 		}
 	}
-
-	// any remaining fields are to be appended
-	for i < len(descInfo.fields) {
-		vali := descInfo.order[i]
-		i++
-		if descInfo.fields[vali].container {
-			bytes = zcode.AppendContainer(bytes, p.vals[vali].Bytes)
+	// Finish building the output by appending the remaining clauses if any.
+	for clause := range rule.append {
```

Comment:
It's wild that existing tests all pass with this loop, but
```yaml
zql: put a=3,b=2,c=1

input: |
  #0:record[a:int32,b:int32]
  0:[1;2;]

output: |
  #0:record[a:int64,b:int64,c:int64]
  0:[3;2;1;]
```
fails with
```diff
        expected and actual outputs differ:
        --- expected
        +++ actual
        @@ -1,3 +1,3 @@
         #0:record[a:int64,b:int64,c:int64]
        -0:[3;2;1;]
        +0:[3;2;3;]
```

It passes with this.
```suggestion
	for _, clause := range rule.append {
```

Reply:
Golly